File: android/src/main/java/com/margelo/nitro/backgroundlocationrelay/LocationRelayEngine.kt
package com.margelo.nitro.backgroundlocationrelay

import android.annotation.SuppressLint
import android.content.Context
import android.location.Location
import android.os.Looper
import com.google.android.gms.location.LocationCallback
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationResult
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.Priority
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

object LocationRelayEngine {
  private const val MIN_INTERVAL_MS = 1_000L

  @Volatile
  private var config: BackgroundLocationRelayConfig? = null

  @Volatile
  private var running = false

  @Volatile
  private var latestLocation: Location? = null

  @Volatile
  private var hasDeliveredInitial = false

  private var locationCallback: LocationCallback? = null
  private var deliveryScheduler: ScheduledExecutorService? = null
  private var deliveryTask: ScheduledFuture<*>? = null

  fun setConfig(context: Context, newConfig: BackgroundLocationRelayConfig) {
    config = newConfig
    ConfigStore.save(context.applicationContext, newConfig)
    RelayLogger.info(ConfigLogger.format(newConfig))
    if (running) {
      RelayLogger.info("Relay is running — applying new configuration.")
      restartLocationUpdates(context.applicationContext)
    }
  }

  fun loadPersistedConfigIfNeeded(context: Context) {
    if (config == null) {
      config = ConfigStore.load(context.applicationContext)
    }
  }

  fun start(context: Context) {
    if (running) {
      return
    }

    loadPersistedConfigIfNeeded(context)

    val currentConfig =
      config ?: throw RelayException("BackgroundLocationRelay is not configured. Call initialize() first.")

    if (!AndroidPermissions.hasLocationPermission(context)) {
      throw RelayException(
        "Location permission (ACCESS_FINE_LOCATION/ACCESS_COARSE_LOCATION) has not been granted. Request it via react-native-permissions before start().",
      )
    }

    running = true
    startLocationUpdates(context.applicationContext, currentConfig)
    RelayLogger.info("Location relay started.")
  }

  fun stop(context: Context) {
    running = false
    latestLocation = null
    hasDeliveredInitial = false
    stopLocationUpdates(context.applicationContext)
    RelayLogger.info("Location relay stopped.")
  }

  fun isRunning(): Boolean = running

  fun restartLocationUpdatesIfRunning(context: Context) {
    if (!running) {
      return
    }

    loadPersistedConfigIfNeeded(context)
    val currentConfig = config ?: return
    startLocationUpdates(context.applicationContext, currentConfig)
  }

  @SuppressLint("MissingPermission")
  fun startLocationUpdates(context: Context, currentConfig: BackgroundLocationRelayConfig) {
    stopLocationUpdates(context)

    val client = LocationServices.getFusedLocationProviderClient(context)
    val request = buildLocationRequest(currentConfig.location)

    val callback =
      object : LocationCallback() {
        override fun onLocationResult(result: LocationResult) {
          val location = result.lastLocation ?: return
          onLocationReceived(location)
        }
      }

    locationCallback = callback
    client.requestLocationUpdates(request, callback, Looper.getMainLooper())
    startDeliveryTimer(currentConfig)
  }

  fun stopLocationUpdates(context: Context) {
    stopDeliveryTimer()

    val callback = locationCallback ?: return
    LocationServices.getFusedLocationProviderClient(context).removeLocationUpdates(callback)
    locationCallback = null
  }

  private fun restartLocationUpdates(context: Context) {
    val currentConfig = config ?: return
    if (!running) {
      return
    }
    startLocationUpdates(context, currentConfig)
  }

  private fun onLocationReceived(location: Location) {
    latestLocation = location

    if (!running) {
      return
    }

    // Deliver the very first fix immediately so consumers do not have to wait a
    // full interval for the initial location; the timer drives every delivery
    // afterwards, including while the device is stationary.
    if (!hasDeliveredInitial) {
      hasDeliveredInitial = true
      deliverLatest()
    }
  }

  private fun startDeliveryTimer(currentConfig: BackgroundLocationRelayConfig) {
    stopDeliveryTimer()

    val intervalMs = currentConfig.location.interval.toLong().coerceAtLeast(MIN_INTERVAL_MS)
    val scheduler = Executors.newSingleThreadScheduledExecutor()
    deliveryScheduler = scheduler
    deliveryTask =
      scheduler.scheduleWithFixedDelay(
        {
          try {
            if (running) {
              deliverLatest()
            }
          } catch (error: Exception) {
            RelayLogger.error("Delivery timer error: ${error.message}")
          }
        },
        intervalMs,
        intervalMs,
        TimeUnit.MILLISECONDS,
      )
  }

  private fun stopDeliveryTimer() {
    deliveryTask?.cancel(false)
    deliveryTask = null
    deliveryScheduler?.shutdownNow()
    deliveryScheduler = null
  }

  private fun deliverLatest() {
    val currentConfig = config ?: return
    val location =
      latestLocation
        ?: run {
          RelayLogger.debug("Skipping delivery: no location fix yet")
          return
        }

    deliver(currentConfig, location)
  }

  private fun deliver(currentConfig: BackgroundLocationRelayConfig, location: Location) {
    Thread {
        try {
          val body = GeoJsonBuilder.buildRequestBody(currentConfig.request.bodyTemplate, location)
          val result =
            HttpClient.post(
              endpoint = currentConfig.request.endpoint,
              headersJson = currentConfig.request.headersJson,
              timeoutMs = currentConfig.request.timeout,
              body = body,
            )

          if (result.success) {
            RelayLogger.debug(
              "Delivered location update (status ${result.statusCode ?: 0})."
            )
          } else {
            RelayLogger.error(
              "Failed to deliver location update: ${result.errorMessage ?: "unknown error"}"
            )
          }
        } catch (error: Exception) {
          RelayLogger.error("Failed to build request body: ${error.message}")
        }
      }
      .start()
  }

  private fun buildLocationRequest(locationConfig: LocationConfig): LocationRequest {
    val builder =
      LocationRequest.Builder(toPriority(locationConfig.accuracy), locationConfig.interval.toLong())

    locationConfig.fastestInterval?.let { builder.setMinUpdateIntervalMillis(it.toLong()) }
    locationConfig.distanceFilter?.let { builder.setMinUpdateDistanceMeters(it.toFloat()) }

    return builder.build()
  }

  private fun toPriority(accuracy: LocationAccuracy?): Int {
    return when (accuracy) {
      LocationAccuracy.LOWEST, LocationAccuracy.LOW -> Priority.PRIORITY_LOW_POWER
      LocationAccuracy.BALANCED -> Priority.PRIORITY_BALANCED_POWER_ACCURACY
      LocationAccuracy.HIGH, LocationAccuracy.HIGHEST, null -> Priority.PRIORITY_HIGH_ACCURACY
    }
  }

  fun hasLocationPermission(context: Context): Boolean =
    AndroidPermissions.hasLocationPermission(context)
}

File: ios/LocationRelayEngine.swift
import CoreLocation
import Foundation

final class LocationRelayEngine: NSObject, CLLocationManagerDelegate {
  static let shared = LocationRelayEngine()

  private static let minIntervalMs: Double = 1_000

  private let locationManager = CLLocationManager()
  private let stateQueue = DispatchQueue(label: "BackgroundLocationRelay.engine")
  private let deliveryQueue = DispatchQueue(label: "BackgroundLocationRelay.delivery")

  private var config: BackgroundLocationRelayConfig?
  private var running = false
  private var latestLocation: CLLocation?
  private var hasDeliveredInitial = false
  private var deliveryTimer: DispatchSourceTimer?

  override private init() {
    super.init()
    locationManager.delegate = self
    locationManager.allowsBackgroundLocationUpdates = false
    locationManager.pausesLocationUpdatesAutomatically = false
    if #available(iOS 11.0, *) {
      locationManager.showsBackgroundLocationIndicator = false
    }
  }

  func setConfig(_ config: BackgroundLocationRelayConfig) {
    stateQueue.sync {
      self.config = config
      ConfigStore.save(config)
    }

    runOnMain {
      self.applyLocationSettings(config.location)
    }

    RelayLogger.info(ConfigLogger.format(config))

    if isRunning() {
      RelayLogger.info("Relay is running — applying new configuration.")
      startDeliveryTimer()
    }
  }

  func loadPersistedConfigIfNeeded() {
    stateQueue.sync {
      if config == nil {
        config = ConfigStore.load()
      }
    }
  }

  func start() throws {
    if isRunning() {
      return
    }

    loadPersistedConfigIfNeeded()

    guard let currentConfig = stateQueue.sync(execute: { config }) else {
      throw RelayError.configurationMissing
    }

    switch locationManager.authorizationStatus {
    case .authorizedAlways:
      break
    case .authorizedWhenInUse:
      RelayLogger.error(
        "Cannot start: background location permission (Always) is required, current status is \(Self.describeAuthorizationStatus(locationManager.authorizationStatus))."
      )
      throw RelayError.backgroundLocationPermissionMissing
    default:
      RelayLogger.error(
        "Cannot start: location permission is \(Self.describeAuthorizationStatus(locationManager.authorizationStatus))."
      )
      throw RelayError.locationPermissionMissing
    }

    stateQueue.sync {
      running = true
      latestLocation = nil
      hasDeliveredInitial = false
    }

    ConfigStore.setWasRunning(true)

    runOnMainSync {
      self.applyLocationSettings(currentConfig.location)
      self.locationManager.startUpdatingLocation()
      // Significant-location-change monitoring keeps the relay alive in the
      // background and lets iOS relaunch the app after it has been terminated.
      if Self.hasAlwaysPermission(for: self.locationManager) {
        self.locationManager.startMonitoringSignificantLocationChanges()
      }
      self.locationManager.requestLocation()
    }

    startDeliveryTimer()

    RelayLogger.info("Location relay started.")
    RelayLogger.info(
      "Location authorization: \(Self.describeAuthorizationStatus(locationManager.authorizationStatus))."
    )
    RelayLogger.info(
      "Location manager settings: pausesAutomatically=\(locationManager.pausesLocationUpdatesAutomatically), distanceFilter=\(locationManager.distanceFilter), allowsBackgroundUpdates=\(locationManager.allowsBackgroundLocationUpdates)."
    )
  }

  func stop() {
    stateQueue.sync {
      running = false
      latestLocation = nil
      hasDeliveredInitial = false
    }

    ConfigStore.setWasRunning(false)
    stopDeliveryTimer()

    runOnMain {
      self.locationManager.stopUpdatingLocation()
      self.locationManager.stopMonitoringSignificantLocationChanges()
    }

    RelayLogger.info("Location relay stopped.")
  }

  func isRunning() -> Bool {
    stateQueue.sync { running }
  }

  /// Resumes tracking after the app has been relaunched by the system (for
  /// example, following a significant-location-change event) without any
  /// JavaScript having run yet.
  @objc func resumeIfNeeded() {
    guard !isRunning() else {
      return
    }

    guard ConfigStore.wasRunning() else {
      return
    }

    loadPersistedConfigIfNeeded()

    guard stateQueue.sync(execute: { config }) != nil else {
      ConfigStore.setWasRunning(false)
      return
    }

    guard Self.hasAlwaysPermission(for: locationManager) else {
      return
    }

    RelayLogger.info("Relaunched by the system — resuming location relay.")
    try? start()
  }

  func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
    RelayLogger.info(
      "Location authorization changed: \(Self.describeAuthorizationStatus(manager.authorizationStatus))."
    )

    runOnMain {
      if Self.hasAlwaysPermission(for: manager) {
        manager.allowsBackgroundLocationUpdates = true
      } else {
        manager.allowsBackgroundLocationUpdates = false
      }
    }

    guard isRunning(), hasLocationPermission() else {
      return
    }

    runOnMain {
      manager.startUpdatingLocation()
      if Self.hasAlwaysPermission(for: manager) {
        manager.startMonitoringSignificantLocationChanges()
      }
    }
  }

  func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
    guard let location = locations.last else {
      RelayLogger.info("Skipping delivery: location update did not contain coordinates.")
      return
    }

    RelayLogger.info(
      String(
        format: "Location update received (%.6f, %.6f, accuracy %.1f m).",
        location.coordinate.latitude,
        location.coordinate.longitude,
        location.horizontalAccuracy
      )
    )

    let shouldDeliverInitial = stateQueue.sync { () -> Bool in
      latestLocation = location
      guard running else {
        return false
      }
      if !hasDeliveredInitial {
        hasDeliveredInitial = true
        return true
      }
      return false
    }

    // Deliver the very first fix immediately so consumers do not have to wait a
    // full interval for the initial location; the timer drives every delivery
    // afterwards, including while the device is stationary.
    if shouldDeliverInitial {
      deliverLatest()
    }
  }

  func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
    RelayLogger.error("Location manager failed: \(error.localizedDescription)")
  }

  private func startDeliveryTimer() {
    guard let currentConfig = stateQueue.sync(execute: { config }) else {
      return
    }

    let intervalMs = max(currentConfig.location.interval, Self.minIntervalMs)
    let interval = intervalMs / 1000

    stopDeliveryTimer()

    let timer = DispatchSource.makeTimerSource(queue: deliveryQueue)
    timer.schedule(deadline: .now() + interval, repeating: interval)
    timer.setEventHandler { [weak self] in
      guard let self = self, self.isRunning() else {
        return
      }
      self.deliverLatest()
    }
    deliveryTimer = timer
    timer.resume()
  }

  private func stopDeliveryTimer() {
    deliveryTimer?.cancel()
    deliveryTimer = nil
  }

  private func deliverLatest() {
    let snapshot = stateQueue.sync { () -> (CLLocation?, BackgroundLocationRelayConfig?) in
      (latestLocation, config)
    }

    guard let currentConfig = snapshot.1 else {
      return
    }

    guard let location = snapshot.0 else {
      RelayLogger.info("Skipping delivery: no location fix yet.")
      return
    }

    deliver(location: location, config: currentConfig)
  }

  private func deliver(location: CLLocation, config: BackgroundLocationRelayConfig) {
    DispatchQueue.global(qos: .utility).async {
      do {
        let body = try GeoJsonBuilder.buildRequestBody(
          bodyTemplate: config.request.bodyTemplate,
          location: location
        )

        let result = HttpClient.post(
          endpoint: config.request.endpoint,
          headersJson: config.request.headersJson,
          timeoutMs: config.request.timeout,
          body: body
        )

        if result.success {
          RelayLogger.info(
            "Delivered location update (status \(result.statusCode ?? 0))."
          )
        } else {
          RelayLogger.error(
            "Failed to deliver location update: \(result.errorMessage ?? "unknown error")"
          )
        }
      } catch {
        RelayLogger.error("Failed to build request body: \(error.localizedDescription)")
      }
    }
  }

  private func applyLocationSettings(_ locationConfig: LocationConfig) {
    locationManager.desiredAccuracy = Self.desiredAccuracy(for: locationConfig.accuracy)
    locationManager.distanceFilter = locationConfig.distanceFilter ?? kCLDistanceFilterNone
    locationManager.pausesLocationUpdatesAutomatically =
      locationConfig.pausesLocationUpdatesAutomatically ?? false

    if #available(iOS 11.0, *) {
      if let showsIndicator = locationConfig.showsBackgroundLocationIndicator {
        locationManager.showsBackgroundLocationIndicator = showsIndicator
      }
    }

    locationManager.allowsBackgroundLocationUpdates =
      Self.hasAlwaysPermission(for: locationManager)
  }

  private func hasLocationPermission() -> Bool {
    switch locationManager.authorizationStatus {
    case .authorizedAlways, .authorizedWhenInUse:
      return true
    default:
      return false
    }
  }

  private static func hasAlwaysPermission(for manager: CLLocationManager) -> Bool {
    manager.authorizationStatus == .authorizedAlways
  }

  private static func describeAuthorizationStatus(_ status: CLAuthorizationStatus) -> String {
    switch status {
    case .notDetermined:
      return "notDetermined"
    case .restricted:
      return "restricted"
    case .denied:
      return "denied"
    case .authorizedAlways:
      return "authorizedAlways"
    case .authorizedWhenInUse:
      return "authorizedWhenInUse"
    @unknown default:
      return "unknown"
    }
  }

  private static func desiredAccuracy(for accuracy: LocationAccuracy?) -> CLLocationAccuracy {
    switch accuracy {
    case .some(.lowest):
      return kCLLocationAccuracyThreeKilometers
    case .some(.low):
      return kCLLocationAccuracyKilometer
    case .some(.balanced):
      return kCLLocationAccuracyHundredMeters
    case .some(.high):
      return kCLLocationAccuracyNearestTenMeters
    case .some(.highest):
      return kCLLocationAccuracyBest
    case .none:
      return kCLLocationAccuracyBest
    }
  }

  private func runOnMain(_ block: @escaping () -> Void) {
    if Thread.isMainThread {
      block()
    } else {
      DispatchQueue.main.async(execute: block)
    }
  }

  private func runOnMainSync(_ block: () -> Void) {
    if Thread.isMainThread {
      block()
    } else {
      DispatchQueue.main.sync(execute: block)
    }
  }
}

@objc(RelayLaunchResumer)
final class RelayLaunchResumer: NSObject {
  @objc static func resume() {
    DispatchQueue.main.async {
      LocationRelayEngine.shared.resumeIfNeeded()
    }
  }
}
